src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataFilter.cs

```diff
@@ -29,21 +29,22 @@ public void OnResourceExecuting(ResourceExecutingContext context)
         /// <inheritdoc />
         public void OnResourceExecuted(ResourceExecutedContext context)
         {
-            _factory.GetTempData(context.HttpContext).Save();
         }
 
         /// <inheritdoc />
         public void OnResultExecuting(ResultExecutingContext context)
         {
+            if (context.Result is IKeepTempDataResult)
+            {
+                _factory.GetTempData(context.HttpContext).Keep();
+            }
+
+            _factory.GetTempData(context.HttpContext).Save();
```

Comment:
The idea is that in a view you're likely buffering and so if we did `OnSendingHeaders` then this change would not be visible to you. There **has** to be a breaking change if we want to support cookies, but that would be the minimal breaking change